### PR TITLE
Pickadate format based on js format

### LINKED ISSUE #237: Date/Time format coherence in pickatime/pickadate
Hi Mathieu,
I noticed hard-coded date/time formatting in kiosks( [1](https://github.com/usnistgov/NEMO/blob/758968e84221424ea648db5ac54a775251c21046/NEMO/apps/kiosk/templates/kiosk/tool_reservation.html#L59-L61), [2](https://github.com/usnistgov/NEMO/blob/758968e84221424ea648db5ac54a775251c21046/NEMO/apps/kiosk/templates/kiosk/tool_reservation.html#L80-L84), [3](https://github.com/usnistgov/NEMO/blob/758968e84221424ea648db5ac54a775251c21046/NEMO/apps/kiosk/templates/kiosk/tool_report_problem.html#L116-L117)) and mobile view ([4](https://github.com/usnistgov/NEMO/blob/758968e84221424ea648db5ac54a775251c21046/NEMO/templates/mobile/new_reservation.html#L72-L74), [5](https://github.com/usnistgov/NEMO/blob/758968e84221424ea648db5ac54a775251c21046/NEMO/templates/mobile/new_reservation.html#L96-L100)).
I was going to submit a PR to make them coherent with all other [Django and Fullcalendar variables](https://github.com/usnistgov/NEMO/blob/758968e84221424ea648db5ac54a775251c21046/resources/settings.py#L63-L79) defined in `settings.py`, but I soon realized that pickadate library uses another formatting for [time](https://amsul.ca/pickadate.js/time/#formatting-rules) and [date](https://amsul.ca/pickadate.js/date/#formatting-rules), which is different from both [Django](https://docs.djangoproject.com/en/3.2/ref/templates/builtins/#std:templatefilter-date) and [FullCalendar](https://fullcalendar.io/docs/v1/formatDate).

How would you suggest to proceed? Shall I add an two additional variables to settings.py?

### PR DESCRIPTION
**Pull request by r-xyz**
Closes #237 by removing hard-coded date/time formatting in kiosks( [1](https://github.com/usnistgov/NEMO/blob/758968e84221424ea648db5ac54a775251c21046/NEMO/apps/kiosk/templates/kiosk/tool_reservation.html#L59-L61), [2](https://github.com/usnistgov/NEMO/blob/758968e84221424ea648db5ac54a775251c21046/NEMO/apps/kiosk/templates/kiosk/tool_reservation.html#L80-L84), [3](https://github.com/usnistgov/NEMO/blob/758968e84221424ea648db5ac54a775251c21046/NEMO/apps/kiosk/templates/kiosk/tool_report_problem.html#L116-L117)) and mobile view ([4](https://github.com/usnistgov/NEMO/blob/758968e84221424ea648db5ac54a775251c21046/NEMO/templates/mobile/new_reservation.html#L72-L74), [5](https://github.com/usnistgov/NEMO/blob/758968e84221424ea648db5ac54a775251c21046/NEMO/templates/mobile/new_reservation.html#L96-L100)).

### CODE REVIEW

**Comment by r-xyz** [comment]
I think this should be enough, but we might add to solve the issue.
The new utilities do:
 - it keeps untouched all format which are the same in js and pickadate
 - it converts all supported formats which are different to the closest equivalent
 - it removes all unsupported formats.

It might not work as expected with very exotic format strings, but it should work in most cases.

**Comment by rptmat57** [comment]
I think I'd like to have the override options in `settings.py`, just to be safe

**Comment by r-xyz** [comment]
Fixed previous approach (I noticed it was not covering all conversions).
Added optional settings in settings.py; if not set, conversion will be attempted.

File: NEMO/utilities.py [hunk 1]
```diff
@@ -74,13 +92,36 @@ def convert_py_format_to_js(string_format: str) -> str:
...
+        return settings.PICK_TIME_FORMAT
+    except AttributeError:
+        string_format = (
+            string_format.replace("%w", "")
+            .replace("%s", "")
+            .replace("%f", "")
+            .replace("%z", "")
+            .replace("%Z", "")
+            .replace("%j", "")
+            .replace(":%S", "")
```
Comment (rptmat57) [inline NEMO/utilities.py:106]:
```suggestion
```

File: NEMO/utilities.py [hunk 2]
```diff
@@ -74,13 +92,36 @@ def convert_py_format_to_js(string_format: str) -> str:
...
+            .replace("%S", "")
+            .replace("%U", "")
+            .replace("%W", "")
+            .replace("%c", "")
+        )
```
Comment (rptmat57) [inline NEMO/utilities.py:111]: might want to add %G, %u, %V, :%z https://docs.python.org/3/library/datetime.html#format-codes

File: NEMO/apps/kiosk/templates/kiosk/tool_reservation.html [hunk 3]
```diff
@@ -56,9 +56,9 @@ <h4>When would you like to reserve the {{ tool }}?</h4>
 	{% for item in tool_reservation_times %}
 	unavailable_times.push([{{ item.start|date:"U" }},{{ item.end|date:"U" }}]);
 	{% endfor %}
-	let date_picker = $('#date').pickadate({format: "dddd, mmmm d", formatSubmit: "yyyy-mm-dd", firstDay: 1, hiddenName: true, onSet: refresh_times});
-	let start_time_picker = $('#start').pickatime({interval: 15, formatSubmit: "H:i", hiddenName: true, formatLabel: format_label});
-	let end_time_picker = $('#end').pickatime({interval: 15, formatSubmit: "H:i", hiddenName: true, formatLabel: format_label});
+	let date_picker = $('#date').pickadate({format: "{{ pick_date_format }}", formatSubmit: "yyyy-mm-dd", firstDay: 1, hiddenName: true, onSet: refresh_times});
+	let start_time_picker = $('#start').pickatime({interval: 15, format: "{{ pick_time_format }}", formatSubmit: "HH:mm", hiddenName: true, formatLabel: format_label});
```
Comment (rptmat57) [inline NEMO/apps/kiosk/templates/kiosk/tool_reservation.html:60]:
```suggestion
	let start_time_picker = $('#start').pickatime({interval: 15, format: "{{ pick_time_format }}", formatSubmit: "HH:i", hiddenName: true, formatLabel: format_label});
```

**Comment by rptmat57** [comment]
looks good, thanks!
would you mind renaming the variables to `pickadate_date_format` and `pickadate_time_format`?

File: NEMO/utilities.py (hunk 1 above)
Comment (rptmat57) [reply NEMO/utilities.py:106]: oh that makes sense. good thinking, let's keep it

File: NEMO/utilities.py [hunk 4]
```diff
@@ -74,13 +92,43 @@ def convert_py_format_to_js(string_format: str) -> str:
...
+            string_format.replace("%w", "")
+            .replace("%s", "")
+            .replace("%f", "")
+            .replace("%z", "")
+            .replace("%Z", "")
+            .replace("%j", "")
+            .replace(":%S", "")
+            .replace("%S", "")
+            .replace("%U", "")
+            .replace("%W", "")
```
Comment (rptmat57) [inline NEMO/utilities.py:109]:
```suggestion
```

File: NEMO/utilities.py [hunk 5]
```diff
@@ -74,13 +92,43 @@ def convert_py_format_to_js(string_format: str) -> str:
...
+            .replace("%c", "")
+            .replace("%U", "")
+            .replace("%W", "")
+            .replace("%G", "")
+            .replace("%U", "")
```
Comment (rptmat57) [inline NEMO/utilities.py:114]:
```suggestion
```

File: NEMO/utilities.py [hunk 6]
```diff
@@ -74,13 +92,43 @@ def convert_py_format_to_js(string_format: str) -> str:
...
+            .replace("%u", "")
+            .replace("%V", "")
+            .replace("%:z", "")
```
Comment (rptmat57) [inline NEMO/utilities.py:117]: this line needs to be before `%z`

File: NEMO/apps/kiosk/templates/kiosk/tool_reservation.html [hunk 7]
```diff
@@ -56,12 +56,12 @@ <h4>When would you like to reserve the {{ tool }}?</h4>
 	{% for item in tool_reservation_times %}
 	unavailable_times.push([{{ item.start|date:"U" }},{{ item.end|date:"U" }}]);
 	{% endfor %}
-	let date_picker = $('#date').pickadate({format: "{{ pick_date_format }}", formatSubmit: "yyyy-mm-dd", firstDay: 1, hiddenName: true, onSet: refresh_times});
-	let start_time_picker = $('#start').pickatime({interval: 15, format: "{{ pick_time_format }}", formatSubmit: "HH:i", hiddenName: true, formatLabel: format_label});
-	let end_time_picker = $('#end').pickatime({interval: 15, format: "{{ pick_time_format }}", formatSubmit: "H:i", hiddenName: true, formatLabel: format_label});
+	let date_picker = $('#date').pickadate({format: "{{ pickadata_date_format }}", formatSubmit: "yyyy-mm-dd", firstDay: 1, hiddenName: true, onSet: refresh_times});
```
Comment (rptmat57) [inline NEMO/apps/kiosk/templates/kiosk/tool_reservation.html:59]:
```suggestion
	let date_picker = $('#date').pickadate({format: "{{ pickadate_date_format }}", formatSubmit: "yyyy-mm-dd", firstDay: 1, hiddenName: true, onSet: refresh_times});
```

File: NEMO/context_processors.py [hunk 8]
```diff
@@ -84,6 +90,8 @@ def base_context(request):
         "time_input_js_format": time_input_js_format,
         "date_input_js_format": date_input_js_format,
         "datetime_input_js_format": datetime_input_js_format,
+        "pickadate_date_format": pickadate_date_format,
+        "pickadate_time_format": pickadatetime_format,
```
Comment (rptmat57) [inline NEMO/context_processors.py:94]:
```suggestion
        "pickadate_time_format": pickadate_time_format,
```

File: NEMO/utilities.py [hunk 9]
```diff
@@ -74,13 +92,43 @@ def convert_py_format_to_js(string_format: str) -> str:
     return string_format
 
 
+def convert_py_to_pick_time(string_format: str) -> str:
+    try:
+        return settings.PICK_TIME_FORMAT
```
Comment (rptmat57) [inline NEMO/utilities.py:97]: the setting name probably needs to be passed as argument, then use `getattr` so you can get the correct one (time or date)

Comment (r-xyz) [reply NEMO/utilities.py]: found a similar approach which actually helps to simplify :)

File: NEMO/apps/kiosk/templates/kiosk/tool_reservation.html [hunk 10]
```diff
@@ -56,12 +56,12 @@ <h4>When would you like to reserve the {{ tool }}?</h4>
 	{% for item in tool_reservation_times %}
 	unavailable_times.push([{{ item.start|date:"U" }},{{ item.end|date:"U" }}]);
 	{% endfor %}
-	let date_picker = $('#date').pickadate({format: "dddd, mmmm d", formatSubmit: "yyyy-mm-dd", firstDay: 1, hiddenName: true, onSet: refresh_times});
-	let start_time_picker = $('#start').pickatime({interval: 15, formatSubmit: "H:i", hiddenName: true, formatLabel: format_label});
-	let end_time_picker = $('#end').pickatime({interval: 15, formatSubmit: "H:i", hiddenName: true, formatLabel: format_label});
+	let date_picker = $('#date').pickadate({format: "{{ pickadate_date_format }}", formatSubmit: "yyyy-mm-dd", firstDay: 1, hiddenName: true, onSet: refresh_times});
+	let start_time_picker = $('#start').pickatime({interval: 15, format: "{{ pickadate_time_format }}", formatSubmit: "HH:i", hiddenName: true, formatLabel: format_label});
```
Comment (rptmat57) [inline NEMO/apps/kiosk/templates/kiosk/tool_reservation.html:60]:
```suggestion
	let start_time_picker = $('#start').pickatime({interval: 15, format: "{{ pickadate_time_format }}", formatSubmit: "H:i", hiddenName: true, formatLabel: format_label});
```

File: NEMO/templates/mobile/new_reservation.html [hunk 11]
```diff
@@ -69,13 +69,13 @@ <h4>When would you like to reserve the {{ item }}?</h4>
 		{% for times in item_reservation_times %}
 		unavailable_times.push([{{ times.start|date:"U" }},{{ times.end|date:"U" }}]);
 		{% endfor %}
-		let date_picker = $('#date').pickadate({format: "dddd, mmmm d", formatSubmit: "yyyy-mm-dd", firstDay: 1, hiddenName: true, onSet: refresh_times});
-		let start_time_picker = $('#start').pickatime({interval: 15, formatSubmit: "H:i", hiddenName: true, formatLabel: format_label});
-		let end_time_picker = $('#end').pickatime({interval: 15, formatSubmit: "H:i", hiddenName: true, formatLabel: format_label});
+		let date_picker = $('#date').pickadate({format: "{{ pickadate_date_format }}", formatSubmit: "yyyy-mm-dd", firstDay: 1, hiddenName: true, onSet: refresh_times});
+		let start_time_picker = $('#start').pickatime({interval: 15,  format: "{{ pickadate_time_format }}", formatSubmit: "H:i", hiddenName: true, formatLabel: format_label });
```
Comment (rptmat57) [inline NEMO/templates/mobile/new_reservation.html:73]:
```suggestion
		let start_time_picker = $('#start').pickatime({interval: 15,  format: "{{ pickadate_time_format }}", formatSubmit: "H:i", hiddenName: true, formatLabel: format_label});
```

File: NEMO/apps/kiosk/templates/kiosk/tool_reservation.html [hunk 12]
```diff
@@ -56,12 +56,12 @@ <h4>When would you like to reserve the {{ tool }}?</h4>
...
-	let date_picker = $('#date').pickadate({format: "dddd, mmmm d", formatSubmit: "yyyy-mm-dd", firstDay: 1, hiddenName: true, onSet: refresh_times});
-	let start_time_picker = $('#start').pickatime({interval: 15, formatSubmit: "H:i", hiddenName: true, formatLabel: format_label});
-	let end_time_picker = $('#end').pickatime({interval: 15, formatSubmit: "H:i", hiddenName: true, formatLabel: format_label});
+	let date_picker = $('#date').pickadate({format: "{{ pickadate_date_format }}", formatSubmit: "yyyy-mm-dd", firstDay: 1, hiddenName: true, onSet: refresh_times});
+	let start_time_picker = $('#start').pickatime({interval: 15, format: "{{ pickadate_time_format }}", formatSubmit: "H:i", hiddenName: true, formatLabel: format_label});
+	let end_time_picker = $('#end').pickatime({interval: 15, format: "{{ pickadate_time_format }}", formatSubmit: "H:i", hiddenName: true, formatLabel: format_label});
 	// set initial date
 	if ('{{ date|default_if_none:'' }}') {
-		date_picker.pickadate('picker').set('select', '{{ date }}', {format: 'yyyy-mm-dd'})
+		date_picker.pickadate('picker').set('select', '{{ date }}', {format: '{{ pickadate_time_format }}'})
```
Comment (rptmat57) [inline NEMO/apps/kiosk/templates/kiosk/tool_reservation.html:64]:
```suggestion
		date_picker.pickadate('picker').set('select', '{{ date }}', {format: '{{ pickadate_date_format }}'})
```

File: NEMO/utilities.py (hunk 9 above)
Comment (rptmat57) [reply NEMO/utilities.py]: that's perfect!

**Pull request merged**